plugins/inventory/gql_inventory.py

```diff
@@ -189,6 +183,44 @@ def add_variable(self, host: str, var: str, var_type: str):
         """
         self.inventory.set_variable(host, var_type, var)
 
+    def create_groups(self, json_data: dict):
+        """Creates groups for host based on `group_by` parameters."""
+        for device in json_data["data"]["devices"]:
+            device_name = device["name"]
+            self.inventory.add_host(device_name)
+            for group_by in self.group_by:
```

Comment:
It was kind of hard to follow the logic within this large, nested for loop block. I think using a "stack" is more readable as to what we're looking to accomplish.

I messed around with it and imo, think this is more readable.

```

    def create_groups(self, device_name, json_data: dict):
        """Creates groups for host based on `group_by` parameters."""
        for device in json_data["data"]["devices"]:
            device_name = device["name"]
            self.inventory.add_host(device_name)
            for group_by in self.group_by:
                # Split dotted group_by's to be found
                parent_attr, *chain, last = group_by.split(".")
                device_attr = device.get(parent_attr)
                # If the parent attr returns None, continue onto next group_by due to attr not existing on device
                if device_attr is None:
                    self.display.display(f"Could not find value for {parent_attr} on device {device_name}")
                    continue

                # If no chain exists, then set the group_name. e.g. tenant.slug
                if not chain:
                    group_name = device_attr

                # Loop over to find the child value if nested. e.g. site.region.slug
                while chain:
                    # We want to pop in reverse order
                    group_name = chain.pop(index=0)

                    if not isinstance(group_name, Mapping):
                        self.display.display(f"Device {device_name} attribute {group_name} is not a dictionary.")
                        continue

                    try:
                        group_name = group_name[group_name]
                    except KeyError:
                        self.display.display(f"Could not find value for {group_name} in {group_by} on device {device_name}")
                        break

                # Now that we have reached our final destination. Let's try and extract the value if a mapping with the following order:
                # 1. The last element of the group_by path
                # 2. The name attribute on the dictionary
                # 3. The slug attribute on the dictionary
                if isinstance(group_name, Mapping):
                    if last in group_name:
                        group_name = group_name[last]
                    elif "name" in group_name:
                        group_name = group_name["name"]
                    elif "slug" in group_name:
                        group_name = group_name["slug"]
                    else:
                        self.display.display(f"No slug value for {group_name} on device {device_name} for group {group_by}.")

                if isinstance(group_name, str):
                    self.inventory.add_group(group_name)
                    self.inventory.add_child(group_name, device_name)
                else:
                    self.display.display(f"{group_name} is not a valid group name for device {device_name} for group {group_by}.")
```

Reply:
I meant to comment as well, however, my thoughts were that this should be split into a few functions instead of one large function

Reply:
I refactored this. I'm personally a fan of the while loop so I went that route. Will leave this open for now.

Reply:
Implemented but changed a little bit!